### PR TITLE
[Unify Tensors PR #1] Replaced pten::Allocation with shared_ptr<memory::Allocation> for Storage

### PR DESCRIPTION
**Pull request by jim19930609**
### PR types
New features

### PR changes
Others
### Describe
Replaced pten::Allocation with shared_ptr<memory::Allocation> for Storage.

### CODE REVIEW

**Review by chenwhql — DISMISSED** [review]
LGTM

File: paddle/pten/core/allocator.h [hunk 1]
```diff
@@ -15,6 +15,7 @@ limitations under the License. */
 #pragma once
```
Comment (Shixiaowei02) [inline paddle/pten/core/allocator.h]: 不建议改这个文件，Tensor 可将接收的 Allocator 从 `pten::Allocator` 改为 `memory::Allocator`。

**Review by Shixiaowei02 — CHANGES_REQUESTED** [review]
除了 `paddle/pten/core/allocator.h`，其它我没问题

File: paddle/pten/core/allocator.h (hunk 1 above)
Comment (jim19930609) [reply paddle/pten/core/allocator.h]: Per offline discussion, replaced AllocateShared() interface with a virtual place() method.

File: paddle/pten/core/allocator.h [hunk 2]
```diff
@@ -15,6 +15,7 @@ limitations under the License. */
...
 
 #include <cstdint>
+#include "paddle/fluid/memory/memory.h"
```
Comment (Shixiaowei02) [inline paddle/pten/core/allocator.h]: 这个头文件建议删除下

**Review by Shixiaowei02 — APPROVED** [review]

**Review by chenwhql — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**